go-init

gregakinman opened the pull request:
@cbrockington 

enables https://github.com/palantir/sls-packaging/issues/164

deadcode is failing because I have a function that's only used in test code, which is allowed in deadcode if you specify `-test`, but it looks like we don't do this. @nmiyake is the way to add arguments to a task to follow the example here? https://github.com/palantir/godel-okgo-asset-errcheck/blob/master/errcheck/config/internal/v0/config.go#L23

Also, it seems like `integration_test/doc.go` exists solely so that there exists at least one non-test source file in the `integration_test` package so that `go install $(./godelw packages)` can succeed, which seems silly - if this is the case, probably better for `./godelw packages` to accept a flag to not produce packages that don't include non-test code (opened https://github.com/palantir/godel/issues/401)

Can I get write access to this repo so I can debug with SSH? Some of the tests rely on `pgrep` and work just fine on my machine but on Circle the calls to `pgrep` exit 1 for some reason...

Commit:
go-init

Commit:
fixing a few linting errors

Commit:
checking errors and undoing rename

Commit:
add back placeholder doc file

Commit:
wow that's dumb

Commit:
style

Commit:
sleeping to try to eliminate flakes

Commit:
bump sleep further?

Commit:
remove sleeps and bump timeout back to original value

iamdanfox commented:
@gregakinman that issue has been dormant since Jan 2017... what's the motivation behind switching from bash -> go now?

(Write access granted, so hopefully CircleCI SSH will work for you now)

iamdanfox commented on init/cli/cli.go:
```diff
@@ -18,29 +18,11 @@ import (
+	app.Usage = "A simple init.sh-style service launcher CLI."
 
-	app.Subcommands = []cli.Command{statusCommand(), startCommand()}
+	app.Subcommands = []cli.Command{startCommand(), statusCommand(), stopCommand()}
```
I think sls-packaging supports a `console` command: https://github.com/palantir/sls-packaging/pull/29

Also, wasn't there functionality to discover git-style subcommand scripts?

gregakinman replied:
I explicitly only want to support `start`, `stop`, and `status`, since those are the only arguments that are in the spec. `search?utf8=✓&q="init.sh+console"&type=` on the end of the internal repo base URL shows that `console` isn't used by the tooling, so it seems fine to remove. We can have a conversation about individual cases for the subcommand scripts if those come up, but I'd want a really good reason to include stuff that's outside of the spec.

iamdanfox replied:
I don't think it's really acceptable to break workflows that people may be relying on without warning, even if they aren't in the spec. If we're converting bash to go, then we need to maintain the [previous behaviour](https://github.com/palantir/sls-packaging/blob/68497f1505f38f40560fec1015ca4871854f559f/gradle-sls-packaging/src/main/resources/init.sh#L163):

```
        echo "Usage: $0 {status|start|stop|console|restart|check${COMMANDS}}"
```

Searching internally, it seems people are relying on `init-reload.sh`.  If we want to get rid of these commands, we could do a release of sls-packaging that deprecates them (and warns people that they shouldn't be used), then do a major release later that removes them.

gregakinman replied:
Yep, would prefer to go the route of "deprecate everything that isn't start/stop/status in sls-packaging then major version bump when we bundle this in sls-packaging instead" rather than "drop-in replacement for the current init.sh without sls-packaging major version bump"

(also obviously we will still have init.sh but it will be a simple pass-through to this binary)

iamdanfox replied:
Ah ok I didn't know whether you wanted to delete the `init.sh` and cutover immediately to this go implementation!  Deprecated then major bump sounds good.

cbrockington commented:
It seems like a lot of the changes in this PR start by just removing a bunch of defaulted command line options.  Is there any particular reason for that?  Their inclusion enables some much simpler end-to-end testing (since you can point your test run to files without having to ensure a bunch of exact locations first), and I'm not sure what we gain by removing them all.

Regardless, if you wish to remove all of the options, could we split this into a "no-op PR" which removes all of the configurability, then a separate PR with all of the additions in functionality (though it's currently unclear to me what these are? which is part of the reason the separation would be nice).

cbrockington commented:
@iamdanfox the reason for the revitalisation of the bash -> go move is to more easily support more complex process creation logic such as the addition of "sidecar processes" in some legacy environments.

gregakinman commented:
@cbrockington I removed the command-line options since, as you noted, it seems like they're there just for testing, and I think in general that adding configurability in production to facilitate testing is a bad thing (it's an artifact of your testing leaking through to production - if there's no clear use case for a property's configurability in production, don't support it if you can help it). Additionally, I can't envision a use case for the flags in production, enforcing a standard for where pidfiles, etc. live is a good thing (and testing that this works when inside that standard setup), and in the current bash world this is the way it is already and it's been fine this whole time. It didn't add much complexity to the tests to set up the SLS distribution structure, so I removed the flags.

It's going to be a fair bit of work to tease apart the changes. I can do it if you really need, but I think it would make it easier for review purposes to look at this PR as completely adding the `init/` directory since it changed so much of what was there. In short, though, the prior state was there was start/status functionality, a few integration tests, and unit tests for start. I added stop and finished adding tests for everything. I also refactored the error reporting to rely on `errors.Wrap` and the `withExitCode` functionality that comes OOTB with the cli package rather than use custom error types with a custom handler.

Also would appreciate a golang-focused review from somebody who knows it well - first time writing Go, so would be good to know if I'm doing anything blatantly wrong/non-idiomatically.

@iamdanfox hmm, I still see this in Circle:
<img width="406" alt="screen shot 2018-06-11 at 15 21 12" src="https://user-images.githubusercontent.com/6173659/41260136-1b99b788-6d8b-11e8-9080-8216f5011945.png">

nmiyake commented:
Sorry, I've been caught in some high-pri work from Friday until now so haven't been able to take a look at this yet. That work has mostly wrapped up so I will review this later today or tomorrow for feedback on content and Golang stuff.

nmiyake commented on init/cli/start.go:
```diff
@@ -15,73 +15,48 @@
-	stdoutFileName := ctx.String(outFileParameter)
+func start(_ cli.Context) error {
+	_, _, err := lib.GetProcessStatus()
+	if err == nil {
```
nit: in a case like this, inline assignment and check:

```go
if _, _, err := lib.GetProcessStatus(); err == nil {
   return nil
}
```

nmiyake commented on init/cli/start.go:
```diff
@@ -15,73 +15,48 @@
-		msg := fmt.Sprintln("Failed to create startup log file", err)
-		return respondError(msg, err, 1)
+		err = errors.Wrap(err, "failed to create startup log file")
+		return cli.WithExitCode(1, err)
```
Now that the code is restructured, you can just inline both these calls (it makes the code more readable/closer to idiomatic Go):

```go
return cli.WithExitCode(1, errors.Wrap(err, ...))
```

nmiyake commented on init/cli/start.go:
```diff
@@ -15,73 +15,48 @@
+Launches the process defined by the static and custom configurations at service/bin/launcher-static.yml and
+var/conf/launcher-custom.yml. Writes its PID to var/run/service.pid and redirects its output to var/log/startup.log. If
+successful, exits 0, otherwise exits 1 and writes an error message to stderr.`,
+		Action: start,
```
Since your actions no longer require the CLI context, I think it's cleaner to have the `Action:` define the function and have the functions themselves be separate functions that don't take a context:

```go
Action: func(_ cli.Context) error {
   return start()
}

func start() error {
    // ...
}
```

If this function were to eventually use flags, then they should be parsed in the `Action` function and the `start` function should receive the arguments as Go types.

gregakinman replied:
nice, didn't know you could do this

nmiyake commented on init/cli/status.go:
```diff
@@ -28,34 +24,20 @@ func statusCommand() cli.Command {
+- 1 if the pidfile exists and can be read but the process is not running
+- 3 if the pidfile does not exist or cannot be read
+If exit code is nonzero, writes an error message to stderr.`,
+		Action: status,
```
Same feedback as for previous command file.

nmiyake commented on init/cli/status.go:
```diff
@@ -28,34 +24,20 @@ func statusCommand() cli.Command {
-		return respondSuccess(isRunning, fmt.Sprintf("Running (%d)\n", pid))
-	case 1:
-		return respondSuccess(isRunning, "Process dead but pidfile exists\n")
+		return cli.WithExitCode(status, err)
```
Inline call and check:

```go
if _, status, err := lib.GetProcessStatus(); err != nil {
```

nmiyake commented on init/cli/stop.go:
```diff
@@ -0,0 +1,60 @@
+import (
+	"github.com/palantir/pkg/cli"
+	"github.com/palantir/go-java-launcher/init/lib"
+	"os"
```
Don't think this passes checks -- imports are not sorted properly. Run `./godelw format` or `./godelw verify` to run the formatting task on this file.

gregakinman replied:
looks like the check isn't catching it:

```
$ ./godelw format
$ echo $?
0
```

gregakinman replied:
oh, the task does the formatting rather than check if it's formatted correctly, i see

nmiyake commented on init/cli/stop.go:
```diff
@@ -0,0 +1,60 @@
+		Usage: `
+Stops the process the PID of which is written to var/run/service.pid. Returns 0 if the process is successfully stopped
+or is not running and the pidfile is removed and returns 1 otherwise.`,
+		Action: stop,
```
Define command function here so `cli.Context` not passed to function

nmiyake commented on init/cli/stop.go:
```diff
@@ -0,0 +1,60 @@
+	// The status tells us more than the error
+	process, status, _ := lib.GetProcessStatus()
+
+	switch status {
```
Inline:

```go
switch process, status, _ := lib.GetProcessStatus(); status {
...
```

gregakinman replied:
_slick_

nmiyake commented on init/cli/stop.go:
```diff
@@ -0,0 +1,60 @@
+	case 3:
+		return nil
+	default:
+		msg := fmt.Sprintf("internal error, process status code not a known value: %d", status)
```
Use `errrors.Errorf(fmtStr, fmtArgs)` rather than `Sprintf` + `errors.New`

nmiyake commented on init/cli/stop.go:
```diff
@@ -0,0 +1,60 @@
+	default:
+		msg := fmt.Sprintf("internal error, process status code not a known value: %d", status)
+		err := errors.New(msg)
+		return cli.WithExitCode(1, err)
```
Inline all:

```go
return cli.WithExitCode(1, errors.Errorf("internal error, ...: %d", status))
```

nmiyake commented on init/lib/start.go:
```diff
@@ -15,32 +15,26 @@
-	err = ioutil.WriteFile(pidFileName, []byte(strconv.Itoa(pid)), 0644)
-	if err != nil {
-		return pid, fmt.Errorf("Failed to write pid file: %s", pidFileName)
+	if err := ioutil.WriteFile(Pidfile, []byte(strconv.Itoa(pid)), 0644); err != nil {
```
Since `pid` is no longer being returned, just do `strconv.Itoa(cmd.Process.Pid)` (no need to store in separate variable)

nmiyake commented on integration_test/go_init_integration_test.go:
```diff
@@ -0,0 +1,228 @@
+
+	unstoppableCommand := "trap '' 15; /bin/echo go-init-testing && /bin/sleep 10000 &"
+	if err := exec.Command("/bin/sh", "-c", unstoppableCommand).Run(); err != nil {
+		panic(err)
```
In tests, I recommend storing error and using `require.NoError` to assert that error is `nil`. So, in this case:

```go
err := exec.Command("/bin/sh", "-c", unstoppableCommand).Run()
require.NoError(t, err, ...)
```

gregakinman replied:
For better readability/transparency when there's an error?

nmiyake replied:
Correct -- `panic` statements can print in an interleaved way with tests when run. This failure mode is much easier to work with. Also has the bonus that it nests a bit less (you don't have to do an `if` check on every call that could produce an error, which is super common in tests)

nmiyake commented on init/lib/test_util.go:
```diff
@@ -0,0 +1,70 @@
+	Setup("testdata/launcher-static-null.yml", "testdata/launcher-custom-null.yml")
+}
+
+func Setup(launcherStaticFilePath string, launcherCustomFilePath string) {
```
For test functions, recommend having them take a `t *testing.T` as the first parameter and doing a `require.NoError` rather than `panic`.

Generally, making test utilities is not recommended -- these functions aren't actually test functions because they're fully exported and callable from any other Go package (and get compiled as part of the non-test program).

gregakinman replied:
> making test utilities is not recommended

You mean "making _public_ test utilities", right? As in, the problem here is that they're exported & published with the main build artifact - it should be fine to have package-private test utilities in a file that ends in `_test.go`

nmiyake replied:
Yes exactly

nmiyake commented on init/lib/stop.go:
```diff
@@ -0,0 +1,55 @@
+func waitForProcessToStop(process *os.Process) error {
+	numSecondsToWait := 240
+	counter := 0
+	for isRunning(process) && counter < numSecondsToWait {
```
It's more idiomatic to use timer and ticker channels here rather than doing an explicit sleep and counter. Here's an implementation in that manner:

```go
func waitForProcessToStop(process *os.Process) error {
	waitDuration := 240 * time.Second
	timer := time.NewTimer(waitDuration)
	defer timer.Stop()
	ticker := time.NewTicker(time.Second)
	defer ticker.Stop()

	for {
		select {
		case <-ticker.C:
			if !isRunning(process) {
				return nil
			}
		case <-timer.C:
			return errors.Errorf("failed to wait for process to stop: process with pid '%d' did not stop within %v", process.Pid, waitDuration)
		}
	}
}
```

nmiyake commented on init/lib/status_test.go:
```diff
@@ -14,28 +14,56 @@
+	assert.Empty(t, process)
+	assert.Equal(t, status, 3)
+	msg := fmt.Sprintf("failed to read pidfile: open %s: no such file or directory", Pidfile)
+	assert.Contains(t, err.Error(), msg)
```
The `assert` and `require` functions take format strings, so no need to create `msg` separately:

```go
assert.Contains(t, err.Error(), "failed to read pidfile: open %s: no such file or directory", Pidfile)
```

gregakinman replied:
hm, I'm not seeing this in practice (with the exact code you listed there):

```
--- FAIL: TestGetProcessStatus_NotRunningPidfileDoesNotExist (0.00s)
        Error Trace:    status_test.go:67
	Error:		"failed to read pidfile: open var/run/service.pid: no such file or directory" does not contain "failed to read pidfile: open %s: no such file or directory"
	Messages:	var/run/service.pid
```

I'll keep debugging, let me know if anything jumps out at you

gregakinman replied:
Yeah, it doesn't look like this is the case unless I'm misreading something:

```
func Contains(t TestingT, s, contains interface{}, msgAndArgs ...interface{}) bool {
	if h, ok := t.(tHelper); ok {
		h.Helper()
	}

	ok, found := includeElement(s, contains)
...
```
I think the error message if it fails can be a format string, but not the actual `contains` since it's untyped

cbrockington replied:
@gregakinman you are correct, the formatting is only for a message to produce on the assertion failing, not on the thing to check, so you were correct the first time.

nmiyake commented on init/lib/status.go:
```diff
@@ -15,50 +15,43 @@
-	if err != nil {
-		return -1, err
+func isRunning(process *os.Process) bool {
+	if err := process.Signal(syscall.Signal(0)); err != nil {
```
Just return comparison directly:

```go
return process.Signal(syscall.Signal(0)) == nil
```

nmiyake reviewed (changes requested):
Thanks for the PR! Left some feedback.

If possible, it would be nice if we could avoid exporting "test" functions -- as currently written, would probably require moving the integration test into the same directory. I don't think that's necessarily a blocker, though, so would work on fixing the other things first.

cbrockington commented:
Sorry, I think my previous message was unclear, I essentially meant that testing _also_ becomes easier with the configuration options, not that testing is the only reason for them.

I think the issue is a difference in perspective of whether this is a tool which "only supports sls-specification and nothing else" or "a tool which performs it's own defined set of tasks, and also the sls-spec".  Personally, I would argue that since this is open-sourced, it has it's own API, and that API includes a bunch of optional configuration, removing that would be a back-compat breaking change which could easily break others workflows.  Maybe we / the project maintainers are fine with that, though I definitely don't think it can be argued it's not part of the API.

Commit:
PR feedback

gregakinman commented:
Agreed that this project has its own API the backwards compatibility of which should be preserved in general (and the optional flags were of course part of it). However, the code that was there before wasn't complete (though it looks like it was published in a couple releases of go-java-launcher (which makes me think maybe we should move go-init to a different repository?), I would consider the go-init part of it pre-release) so the chance somebody was using it is basically nil. We won't be breaking any workflows, so I'm fine with making the change and would prefer to.

iamdanfox commented on init/cli/stop.go:
```diff
@@ -0,0 +1,59 @@
+		Name: "stop",
+		Usage: `
+Stops the process the PID of which is written to var/run/service.pid. Returns 0 if the process is successfully stopped
+or is not running and the pidfile is removed and returns 1 otherwise.`,
```
Documentation should also include how long this will wait for a process to exit before it is considered 'failed to stop'

iamdanfox commented:
I'm really happy that this means we're actually gonna have tests for the init script!  Also it opens the door to do some retention on the startup log so that it can't grow in an unbounded way (cc @ellisjoe).

gregakinman commented:
Yep! Way better than the way we tested our bash... https://github.com/palantir/sls-packaging/pull/29#issuecomment-164883865 😂

Commit:
more feedback/fixing things

Commit:
fixing format string issues

Commit:
fix type bug

gregakinman commented:
Ok, I think all of the PR feedback was addressed, lmk if anyone has more comments. ~~Sadly still can't SSH :(~~ of course logging out and back in worked

Commit:
serialize tests that have side effects

Commit:
don't run integration tests twice

gregakinman commented:
any merge+release objections?

nmiyake commented:
Sorry haven't had a chance to review the modifications -- can commit to doing so before tomorrow.

nmiyake commented on init/cli/start.go:
```diff
@@ -29,59 +28,35 @@ func startCommand() cli.Command {
-	if err != nil {
-		msg := fmt.Sprintln("Failed to start process", err)
-		return respondError(msg, err, 1)
+	if err = lib.StartCommand(cmd, outputFile); err != nil {
```
Minor, but when using this style, do a `err :=` assignment -- this ensures that the variable is scoped just to this block of code (and doesn't update the value of the outer `err` variable)

nmiyake reviewed (commented):
Very minor feedback, but otherwise LGTM

nmiyake commented:
Content looks good to me on the whole -- made a very minor comment, but shouldn't blcok merging/approval. I don't have merge permissions on the repo, though, so I can't actually do the merge.

Commit:
small nit
